### ISSUE TITLE
Hook remove not work

### ISSUE THREAD

**Issue by glingyan** [pytorch/pytorch#36372]
## 🐛 Bug

when backward(create_graph=True) , remove hook handle not work,
## To Reproduce
run the test code

import torch
from torch import nn
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 32, kernel_size=5)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=5)
        self.pool = nn.MaxPool2d(2, 2)
        self.relu = nn.ReLU()
        self.linear = nn.Linear(64 * 5 * 5, 10)

    def forward(self, input):
        conv1 = self.conv1(input)
        pool1 = self.pool(conv1)
        self.relu1 = self.relu(pool1)
        conv2 = self.conv2(self.relu1)
        pool2 = self.pool(conv2)
        self.relu2 = self.relu(pool2)
        relu2 = self.relu2.view(self.relu2.size(0), -1)
        return self.linear(relu2)
conv_inputs = []
conv_inputs_grad = []

def fw_hook_fn(module, input, output):
        print("fw hook {}".format(type(module).__name__))
        conv_inputs.append(input)

def bwd_hook_fn(module, grad, output):
    print("bwd hook {}".format(type(module).__name__))
    conv_inputs_grad.append(grad)

model = Net()
fwd_handle = model.conv1.register_forward_hook(fw_hook_fn)
bwd_handle = model.conv1.register_backward_hook(bwd_hook_fn)

# Random input and labels; we're not really learning anything
input = torch.rand(1, 3, 32, 32)
label = torch.randint(0, 10, (1,))
input.requires_grad = True
output = model(input)
loss = nn.CrossEntropyLoss()(output, label)
loss.backward(create_graph=True)

fwd_handle.remove()
bwd_handle.remove()
output = model(input)
#still print gradient
loss.backward()

1. run above code , it will output
fw hook Conv2d
bwd hook Conv2d
bwd hook Conv2d
## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by glingyan** [comment]
update the python code as file
[faile_to_remove_hook.txt](https://github.com/pytorch/pytorch/files/4460444/faile_to_remove_hook.txt)

**Comment by glingyan** [comment]
i think it is the same with https://github.com/pytorch/pytorch/issues/25723
